### ISSUE TITLE
validate_subset raising when enum includes negative integers

### ISSUE THREAD

**Issue by viniciussbs** [elixir-ecto/ecto#4445]
### Elixir version

1.15.4

### Database and Version

PostgreSQL 14

### Ecto Versions

ecto 3.10.3

### Database Adapter and Versions (postgrex, myxql, etc)

postgrex 0.17.5

### Current behavior

Hi!

I'm trying to validate a field with `validate_subset(changeset, :foo, [-1, 0])` and this is raising: 

```
** (UnicodeConversionError) invalid code point -1
```

The `changes` being validated: `%{foo: [1]}`.

The same thing happens if I use other negative integers instead of `-1`.

### Expected behavior

I expect it to mark the changeset as invalid and add the proper default validation message to the field `:foo`.

**Comment by josevalim** [comment]
What is the type of the foo column? String? Number?

**Comment by viniciussbs** [comment]
It's `{:array, :integer}`, no default value.

**Comment by josevalim** [comment]
Thanks. We might be failing to build the error message then. Do you have the whole stacktrace for the error above?

**Comment by viniciussbs** [comment]
```
     ** (UnicodeConversionError) invalid code point -1
     code: assert %{foo: ["has an invalid entry"]} == errors_on(changeset)
     stacktrace:
       (elixir 1.15.4) lib/list.ex:1103: List.to_string/1
       (knowtrex 1.3.0) test/support/data_case.ex:63: anonymous fn/2 in Knowtrex.DataCase.errors_on/1
       (elixir 1.15.4) lib/enum.ex:2510: Enum."-reduce/3-lists^foldl/2-0-"/3
       (ecto 3.10.3) lib/ecto/changeset.ex:3744: anonymous fn/3 in Ecto.Changeset.merge_keyword_keys/3
       (elixir 1.15.4) lib/enum.ex:2510: Enum."-reduce/3-lists^foldl/2-0-"/3
       (ecto 3.10.3) lib/ecto/changeset.ex:3738: Ecto.Changeset.traverse_errors/2
```

**Comment by josevalim** [comment]
So this is not really an Ecto issue. The `errors_on` function is interpolating the subset into the message, trying to build it, but `[0, -1]` is not valid. I recommend instead passing a custom `:message` to `validate_subset`. We should probably consider changing the default value here.

**Comment by viniciussbs** [comment]
Good catch! Sorry, I've missed that one line in the stacktrace.

So, the issue is indeed in this `errors_on` implementation. It's traversing the changeset errors to build a nice message, so the subset is interpoleted and `to_string/1` is called. Using `inspect/1` fixes the error but I've noticed that we already have another implementation using `Gettext` in the code base, so I've used and it works, so I'll keep only one implementation.

I have a question, though: It's expected `to_string([-1])` to fail? I think so because the docs says that the conversion to string follows `String.Chars` protocol, right? Just to double check.
